### PR TITLE
Add node_module symlink support

### PR DESCRIPTION
**Pull request by mlabrum**
Issue:
If you have a symlink'd node_module then react-native adds that as a project root, as storybook is manually setting the project roots, storybook will need to detect the symlink and add it as a root

I've ran into this issue because we have a monorepo with symlink'd dependencies....

## What I did
Edited the storybook start command to call the react-native symlink detection script and add those to the project root

## How to test
- Move a node module to another location
- Create a symlink to that node_module
- Ensure that React Native has picked that module up

Thanks!

### CODE REVIEW

**Comment by tmeasday** [comment]
Could this solve some of the problems we've been having too? https://github.com/storybooks/storybook/issues/1301

**Comment by mlabrum** [comment]
It could, we're pretty much doing the same thing

```
packages/app
  - node_modules
      - common -> symlink to ../../common via npm5 file:

packages/common

packages/web
  - node_modules
      - common -> symlink to ../../common via npm5 file:
```

If you're using windows you'll need to do the following:
- Upgrade watchman to a version that supports symlinks on windows
- Patch react native https://github.com/facebook/watchman/issues/479 to support the new watchman paths

If you're using linux/mac the symlink should just work once the packager has the new root.

I've tested this PR and it does solve my problem.

**Comment by ndelangen** [comment]
@mlabrum Thank you so much for this!

Another stepping stone to getting storybook development in a amazing state!

If you could have a look at the linting or add a `eslint-ignore-next-line` or something we can get this merged I think.

**Comment by mlabrum** [comment]
Hey,

I'm glad I could give back :) I'll fix up that linting error

Cheers,

**Comment by mlabrum** [comment]
Hey, I can't seem to be able to see the linting results from circle CI, The linting section just seems to spin forever

**Comment by ajwhite** [comment]
This is quite clever. I think many of us run into this problem. 

I, in particular, use Haul to gain the benefit of webpack. But having this support-built in is a nice frictionless experience.

Can you think of any scenarios where this might cause problems, as the packager would normally skip symlinks? Wondering if it should be a configuration option that users can opt in or out of.

**Comment by mlabrum** [comment]
@ajwhite this solution pretty much mocks what react native does when it loads its config
https://github.com/facebook/react-native/blob/113e046444e090e38f89574620210ba86b711fed/local-cli/util/Config.js#L153
but it seems by storybook providing its own project root and directories this is ignored within react native.... so storybook has to add these directories.

I'm not sure how Haul would handle these extra project roots....but react natives packager will be fine

**Comment by ndelangen** [comment]
@ajwhite Do you think you could test this for Haul and review & merge?

**Comment by tmeasday** [comment]
I just took a look at this and it works great!

With this change we can remove the RN apps from our list of "non-lerna-ed" things. Perhaps in a follow up PR.

**Comment by ndelangen** [comment]
> With this change we can remove the RN apps from our list of "non-lerna-ed" things. Perhaps in a follow up PR.

HOLY SMOKES, I'd say that calls for some 🎊 🎈 🎉 💯

**Review by ndelangen — APPROVED** [review]

**Comment by tmeasday** [comment]
Don't get the 🍾 out just yet, bootstrapping was creaking a bit when I tried to test it, but I confirmed that the basic problem that caused us to do that was solved by this change.

**Comment by Gongreg** [comment]
I just have a question:
Is it safe to depend on react-native/local-cli/util/findSymlinksPaths? This is a quite deep import.

File: app/react-native/src/bin/storybook-start.js [hunk 1]
```diff
@@ -41,8 +41,17 @@ server.listen(...listenAddr, err => {
...
+  let symlinks = []
+  
+  try{
+    const findSymlinksPaths = require('react-native/local-cli/util/findSymlinksPaths'); // eslint-disable-line global-require
+    symlinks = findSymlinksPaths(path.join(projectDir, 'node_modules'), [projectDir]);
+  }catch(e){
+   console.warn(`Unable to load findSymlinksPaths: ${e.message}`); 
+  }
 
+  const projectRoots = (configDir === projectDir ? [configDir] : [configDir, projectDir]).concat(symlinks);
```
Comment (danielduan) [inline app/react-native/src/bin/storybook-start.js]: I think it'd be a good option somewhere to be able to specify a different project root through `storybook-start` for those who have components and assets that need to be included from outside of the storybook root project.

**Review by danielduan — APPROVED** [review]
Looks good!

File: app/react-native/src/bin/storybook-start.js (hunk 1 above)
Comment (mlabrum) [inline app/react-native/src/bin/storybook-start.js]: I agree, it would be nice to have, but probably belongs in a different PR

**Comment by mlabrum** [comment]
@Gongreg It's as safe as it could be, I've wrapped it in a try/catch...the only option around this would be to copy their code into the repo.

it's probably the same level of risk as depending on `node_modules/react-native/local-cli/cli.js`

**Review by Gongreg — APPROVED** [review]

**Comment by ndelangen** [comment]
@tmeasday Are you ready to approve this?

**Comment by tmeasday** [comment]
Yes!

**Review by tmeasday — APPROVED** [review]

**Comment by ndelangen** [comment]
Thank you @mlabrum for this PR!
